### DISCUSSION TITLE
About video seamless loop?

### DISCUSSION THREAD

**Discussion by Jason0247** [airsdk/Adobe-Runtime-Support#1316] [Q&A]
When air uses stagevideo or video, when it is detected that the video is playing to the end of the film, it will start playing again in the seek (0) mode, but the video will get stuck during replay, so it is impossible to realize seamless circular playback.

I would like to ask you if there is any way to play seamlessly during the cycle, or if you have any opinions on the format coding of the video?

thank you

**Comment by ajwfrost** [comment]
Hi

We did some updates a while back to try to resolve this problem with seek(0). I thought the only challenges we faced were some glitches that meant that it wasn't completely smooth around the first half-second or so, but are you saying it's sticking completely? and is this within the replay (e.g. time 1s+) or at the point of the reset/seek?

There is a slight challenge with seek(0) because it would mean the decoder pipeline needs to be refilled again with the content and wouldn't play until enough content is ready and buffered. We're looking specifically at seamless transitions between streams, and at smoothly repeating a track or set of tracks, during our review of the multimedia that's ongoing currently. So I would hope that this sort of issue can be made a lot simpler in the future..

One workaround I know a company has done in order to ensure smooth transitions is to actually have two different NetStream/Video pairs and have the second one ready to roll but invisible, then when the first one finishes this second one is then unpaused and will immediately start playing.

thanks

**Comment by Jason0247** [comment]
Yes, when making a video, the head and tail are the same picture. I want to switch seamlessly, but I have to buffer every time. There will be a phenomenon of non fluency and jamming.

I also tried to solve it with two videos, but the effect is still average. Because the video like this seamless loop will be obvious in 0.5 seconds.

**Reply by VCone** [reply]
Hi, have you tried putting your video into an FLV container and then using the **appendBytes** API to constantly re-feed the byte array of same video data (per each _"playback ending"_ event)? That should loop the video seamlessly.

**Reply by Jason0247** [reply]
Hi, I haven't tried this way before. Do you have to be flv? Is there any example code you can refer to? thank you!

**Reply by ajwfrost** [reply]
That's quite a good idea .. we used this mechanism in the past to implement RTSP support for a customer (to work on a normal Flash Player installation). No example code I'm afraid but you need to wrap the A/V packets into the FLV container format (and add the header) and send this into appendBytes. It supports H.264/AAC that way too...

FLV format = https://www.adobe.com/content/dam/acom/en/devnet/flv/video_file_format_spec_v10.pdf (ignore the F4V stuff in there...)

**Comment by xpojosh** [comment]
@ajwfrost  -- I know you guys are working on a whole new multimedia pipeline, so I have been sitting on my hands vis-a-vis our various video woes.  However, I saw that in your latest release you've been working on the data generation mode of netstreams, so I thought I would ping you on something we've observed:

As some other user have done, I have been trying to create seamless movie loops by re-containerizing .mp4 files as .flv's and then appending bytes to the net stream, using the NetStreamAppendBytesAction.RESET_BEGIN flag when reaching the loop point. I noticed that we get 1 frame of video from the next iteration being displayed at random points just prior to the first movie iteration completing.  To understand better what is happening, I used render textures tiled out, each 1 frame earlier the next to catch the issue (read like a book). When onMeta data of the next movie iteration is called, I pause the video. Here is what I get:

![harmanLoop](https://github.com/airsdk/Adobe-Runtime-Support/assets/76958758/b60d9542-259d-4bb7-907e-dc7ef03cd330)

Frame #7 of the brown carpeted hallway is actually the first frame of beginning of the movie. Every time I run this, the exact location of this frame will move around, happening as early as frame 3. 

As a sanity check, I used the same source code to compile a version using Adobe version 31 of the runtime. It consistently displays the correct sequence of frames as such:

![adobeLoop](https://github.com/airsdk/Adobe-Runtime-Support/assets/76958758/84c4f3e7-b31b-4028-b602-7c1afb4b6e01)

I've tested this with various mp4s, using ffmpeg to re-containerize the movie, while simply copying over the codec. Based on the fact that the location of the errant frame keeps moving makes this feel like some sort of race condition in the harman code.

Any thoughts?

Thanks!

BTW I know that @itlancer has been working on this approach. I don't suppose he has seen something similar?

**Reply by ajwfrost** [reply]
Hi

Thanks for this analysis .. it looks like there is some out-dated frame being picked up somehow. I don't suppose it's possible to get a test case for this which we could run that shows that output/glitch?

thanks

**Comment by xpojosh** [comment]
Sorry about the delay! Here is a simple test case. It uses a video which counts up towards 30 seconds (actually the last frame will show 29.27 with 897 frames decoded). Attached you will find the movie file. 

To simplify this project I have not added the rendertexture snapshots -- however, if you take a slow motion video of the output on your phone you can see it quickly jump to the first frame of the next video and then back to the last few frames of the video.

I've added a "loopVideo" flag in the constructor. If you set this to false, the video will only play once -- however, you will also notice that it doesn't get to the very last frame, but rather stops showing the video at 29:05. The video needs additional bytes added to the the net stream in order to actually get to the end o the video. I assume this could be related to our issue.

[fmst.flv.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/15152176/fmst.flv.zip)


```
package com.mydomain.video
 {
	import com.adobe.utils.AGALMiniAssembler;

	import flash.desktop.NativeApplication;
	import flash.display.Sprite;
	import flash.display3D.Context3D;
	import flash.display3D.Context3DProgramType;
	import flash.display3D.Context3DVertexBufferFormat;
	import flash.display3D.IndexBuffer3D;
	import flash.display3D.Program3D;
	import flash.display3D.VertexBuffer3D;
	import flash.display3D.textures.VideoTexture;
	import flash.events.Event;
	import flash.events.InvokeEvent;
	import flash.events.NetStatusEvent;
	import flash.events.SecurityErrorEvent;
	import flash.filesystem.File;
	import flash.filesystem.FileMode;
	import flash.filesystem.FileStream;
	import flash.geom.Matrix3D;
	import flash.net.NetConnection;
	import flash.net.NetStream;
	import flash.net.NetStreamAppendBytesAction;
	import flash.utils.ByteArray;
	
	public class VideoLoopTest extends Sprite {
		
		private const movieWidth : int = 480;
		private const movieHeight : int = 360;
		private var context3D:Context3D;
		private var indexbuffer:IndexBuffer3D;
		private var matrix3D:Matrix3D = new Matrix3D();
		private var nc:NetConnection;
		private var ns:NetStream;
		private var videoTexture:VideoTexture;
		private var videoFile : File;
		private var logger : PipeLogger;
		private var fileLength:uint =0;
		private var fileStream:FileStream;
		private var currFilePosition:uint =0;
		private var _byteArray:ByteArray;
		private var loopVideo:Boolean = false;
		private var firstPlay:Boolean = true;
		private var burnRate:uint;
		private var previousAddTime:Number = 0;
		
		public function VideoLoopTest() {
			
			loopVideo = true;
						
			addEventListener(Event.ADDED_TO_STAGE, init);
			NativeApplication.nativeApplication.addEventListener(InvokeEvent.INVOKE, onInvoke);
			
			videoFile = new File("C:\\Users\\yourUsername\\");
			videoFile = videoFile.resolvePath("fmst.flv");
			
			logger = new PipeLogger();
			logger.localLog("Loop test. Playing video: " + videoFile.url);	
			
		}
		
			protected function onInvoke(event : InvokeEvent) : void {
			
			logger.localLog("onInvoke called. Setup native application" );
		
			NativeApplication.nativeApplication.activeWindow.width = movieWidth*2;
			NativeApplication.nativeApplication.activeWindow.height = movieHeight*2;
			NativeApplication.nativeApplication.activeWindow.x = 0;
			NativeApplication.nativeApplication.activeWindow.y = 0;
		}
		
		private function init(e:Event):void {
			removeEventListener(Event.ADDED_TO_STAGE, init);
			
			fileLength = videoFile.size;
			fileStream = new FileStream();
			fileStream.open(videoFile, FileMode.READ);
			_byteArray = new ByteArray;
			fileStream.readBytes(_byteArray, 0, 5000);
			currFilePosition = 5000;
			

			if (stage.stage3Ds.length > 0){
			stage.stage3Ds[0].addEventListener(Event.CONTEXT3D_CREATE, contextCreated);
			stage.stage3Ds[0].requestContext3D();


			} else {
				logger.localLog("Error: there is no Stage3D available");
			}
		}
		
		private function contextCreated(event:Event):void {
			
			context3D = stage.stage3Ds[0].context3D;
			context3D.enableErrorChecking = true;
			context3D.configureBackBuffer(movieWidth * 2, movieHeight * 2, 0, true);
			logger.localLog(context3D.driverInfo);
			
			
			var vertices:Vector.<Number> = Vector.<Number>([
				0.5, -1, 0, 1, 0,
				0.5, 0.5, 0, 1, 1,
				-1, 0.5, 0, 0, 1,
				-1,-1, 0, 0, 0
			]);
		
			// create the buffer to upload the vertices
			var vertexbuffer:VertexBuffer3D = context3D.createVertexBuffer(4, 5);
			// upload the vertices
			vertexbuffer.uploadFromVector(vertices, 0, 4);
			// create the buffer to upload the indices
			indexbuffer = context3D.createIndexBuffer(6);
			// upload the indices
			indexbuffer.uploadFromVector(Vector.<uint>([0, 1, 2, 2, 3, 0]), 0, 6);
			
			// create the mini assembler
			var vertexShaderAssembler:AGALMiniAssembler = new AGALMiniAssembler();
			// assemble the vertex shader
			vertexShaderAssembler.assemble(Context3DProgramType.VERTEX, "m44 op, va0, vc0\n" + "mov v0, va1");

			var fragmentShaderAssembler:AGALMiniAssembler = new AGALMiniAssembler();
			// assemble the fragment shader
			fragmentShaderAssembler.assemble(Context3DProgramType.FRAGMENT, "tex ft1, v0, fs0 <2d,linear, nomip>\n" + "mov oc, ft1");
			// create the shader program
			var program:Program3D = context3D.createProgram();
			// upload the vertex and fragment shaders
			program.upload(vertexShaderAssembler.agalcode, fragmentShaderAssembler.agalcode);
			
			// set the vertex buffer
			context3D.setVertexBufferAt(0, vertexbuffer, 0, Context3DVertexBufferFormat.FLOAT_3);
			context3D.setVertexBufferAt(1, vertexbuffer, 3, Context3DVertexBufferFormat.FLOAT_2);
			
			videoTexture = context3D.createVideoTexture();
			context3D.setTextureAt(0, videoTexture);
			
			context3D.setProgram(program);
			matrix3D.appendScale(1, -1, 1);
			context3D.setProgramConstantsFromMatrix(Context3DProgramType.VERTEX, 0, matrix3D, true);
			
			nc = new NetConnection();
			nc.addEventListener(NetStatusEvent.NET_STATUS, netStatusHandler);
			nc.addEventListener(SecurityErrorEvent.SECURITY_ERROR, securityErrorHandler);
			nc.connect(null);
		}
		
		private function netStatusHandler(event:NetStatusEvent):void {
			logger.localLog(event.info.code);
			switch (event.info.code){
				case "NetConnection.Connect.Success":
					ns = new NetStream(nc);
					ns.addEventListener(NetStatusEvent.NET_STATUS, netStatusHandler);
					ns.client = {onMetaData:getMeta};
					videoTexture.attachNetStream(ns);
					videoTexture.addEventListener(Event.TEXTURE_READY, renderState);
										
					ns.play(null);
					ns.appendBytes(_byteArray);

				break;
				case "NetStream.Play.StreamNotFound":
					logger.localLog("Stream not found");
				break;
				case "NetStream.Play.Stop":
					logger.localLog("End of video reached");
				break;
			}
		}
		
		//Metadata handler
		private function getMeta(mdata:Object):void {
			logger.localLog("Meta data. Total play time= " + mdata.duration);
			
					if (firstPlay) {
					burnRate = fileLength / mdata.duration;
					_byteArray = new ByteArray();
					
					logger.localLog("Our burn rate = " + burnRate);
					fileStream.readBytes(_byteArray, 0, 2*burnRate);
					currFilePosition = 2*burnRate + currFilePosition;
					_byteArray.position = 0;
					
					logger.localLog("Appending bytes. Bytes available " + _byteArray.bytesAvailable);
				
					ns.appendBytes(_byteArray);
					addEventListener(Event.ENTER_FRAME, onEnterFrame);
					previousAddTime = 0;
					firstPlay = false;
				} else {
					// We've looped	
				}
			
		}
		
		
	private function onEnterFrame(event : Event) : void {

			
		var elapsedTime:Number = (ns.time - previousAddTime);
			
		if (elapsedTime >1 || ns.bufferLength < 1) {
		
		logger.localLog("Current buffer length in seconds = " + ns.bufferLength);
			
			_byteArray = new ByteArray();
			
			previousAddTime = ns.time;
			var neededBytes:uint;
			
			if (ns.bufferLength < 1) {
			// let's increase the buffer by a second	
			neededBytes = burnRate;
			} else {
			 neededBytes = elapsedTime * burnRate;
			}
			
			if ((currFilePosition + neededBytes) >= fileLength) {
			fileStream.readBytes(_byteArray, 0, (fileLength - currFilePosition));		
			_byteArray.position = 0;
			
			neededBytes -= fileLength - currFilePosition;
			
			currFilePosition = 0;
			fileStream.position = 0;

			logger.localLog("We are at the end!  Appending bytes available " + _byteArray.bytesAvailable);	
			ns.appendBytes(_byteArray);
			

			if (!loopVideo) {
				
			logger.localLog("Our flv movie does not need to play again. Let us stop here ");
			removeEventListener(Event.ENTER_FRAME, onEnterFrame);
	
			} else {
			logger.localLog("We will be looping. Update flv header ");	
			ns.appendBytesAction(NetStreamAppendBytesAction.RESET_BEGIN);	
			}	
			}
					
			// Any more bytes needed for this iteration?		
			if (neededBytes >0) {				
				
			_byteArray = new ByteArray();	
			fileStream.readBytes(_byteArray, 0, neededBytes);

			
			_byteArray.position = 0;
			currFilePosition += neededBytes;		

			logger.localLog("Appending bytes. Bytes available " + _byteArray.bytesAvailable);
			ns.appendBytes(_byteArray);
			}
			
		}
		
			
		
		}

		private function securityErrorHandler(event:SecurityErrorEvent):void {
			trace("securityErrorHandler:", event.text);
		}
		
		private function renderState(e:Event):void {
			render();
		}

		private function render(event:Event = null):void {
			context3D.clear(0, 0, 0, 1);
			context3D.drawTriangles(indexbuffer);
			context3D.present();
		}
		
		
	}
}
```

**Reply by ajwfrost** [reply]
Great, thanks for this, and we have it reproducing here .. it seems we get frame 893, then 000, then 896, 897, 001...
And if we don't loop, we stop at frame 869, although it's noticeable that the video stops before the audio.

Will look into both of these ones...

**Reply by ajwfrost** [reply]
Hi again

So there are two separate errors happening with this...

The first (simpler) one was if we don't loop, the video finishes a second or so early. This is because the Windows decoder was telling us it needed more input, which of course it wasn't getting. But, that's because it didn't know the stream had ended... There was a bit of an internal misconnection in the 'EOF' notifications, but we've corrected this so that we can let the decoder know to keep processing all the data its got and to run its buffer dry. But, to let it know the stream has actually ended, you need to send in the "end sequence" action, and then make sure you don't pass in any more bytes: from your test case, where you remove the enter frame listener, you would need to add:
```
	ns.appendBytesAction(NetStreamAppendBytesAction.END_SEQUENCE);	
	neededBytes = 0;
```

The second one appears to be due to the fact that the stream receives the reset event and then the FLV data includes the SPS/PPS information from the H.264 stream - but for that, we don't have a timestamp set up, and this appears to confuse the Windows media foundation decoder. Switching to use the cached "current" timestamp then resolves this issue.

The only drawback you'll find with this approach for looping - where you reset the time to say there will be a timestamp jump backwards - is that we don't then know the intended timing of the first frame that is sent in again. The code just uses the "last timestamp" from the previously received data. So basically, frame 896 is at time 29.950, frame 897 is at time 29.983, then we get a reset, then frame 0 arrives with timestamp 0 and we adjust that to be the same as the previous one i.e. 29.983. But because frames 897 and 0 both have the same timestamp, only the more recent one is actually displayed.. so you would never spot "897" appearing on the screen.

I guess it would be possible to send in a manually-crafted video frame with an updated timestamp (30.016) but no/irrelevant content, to ensure the decoder still displays that last frame 897 and then displays the new frame 0 at 30.016..?

Anyway -> we'll look at getting this into a release so that you can check it out..

thanks

**Reply by xpojosh** [reply]
@ajwfrost 

My apologies for the delayed comment. I forgot to post this earlier...

You had mentioned that missing frames at the end of the movie could be rectified with the by setting the NetStreamAppendBytesAction.END_SEQUENCE at the end of the movie. I had not gone down that path originally as we had previously seen the same behavior with a standard netstream.play() call (before even toying with the appendBytes approach) and assumed that there were other outstanding issues in the Harman video pipeline.

If you take the test case and instead use:

					videoTexture.attachNetStream(ns);
					videoTexture.addEventListener(Event.TEXTURE_READY, renderState);
					ns.play(videoFile.url);	

you will see that video does not make it to the end (getting only to 29:05). This problem does not exist in the Adobe runtime.

Thanks kindly!

Josh

**Reply by ajwfrost** [reply]
Hi @xpojosh 

Sorry, I wasn't clear enough for this:
> You had mentioned that missing frames at the end of the movie could be rectified with the by setting the NetStreamAppendBytesAction.END_SEQUENCE at the end of the movie

Actually, what I meant was, with a fix that we can make to the AIR runtime code, then if you send in the end sequence command, we can put the decoder into its "drain" mode and it will output all the frames... So, yes, you would still find it's not using them up properly. The actual code issue was just an incorrect override in a class which meant the end-of-stream function wasn't called properly, because it didn't have the correct function signature...

So, please check after we've issued the next release.. which should be towards the end of this month..

thanks

**Reply by xpojosh** [reply]
@ajwfrost  --- I just tried out the new 51.0.1.2 release. Both the looping functionality with appendBytes and a normal play once both work like a champ, getting to the end of the movie. Thanks so much!

I hope I don't sound too greedy, but I don't suppose you are going to add support for appendBytes with mp4 files as was discussed here? https://github.com/airsdk/ANE-PlayAssetDelivery/issues/20

Thanks again for all your great work!
josh

**Reply by ajwfrost** [reply]
Hi @xpojosh  - thanks for the feedback on those, glad it worked for you too!

If I recall correctly, we are now able to accept an MP4 file if it's provided all at once as a byte array (or IDataInput). So it's not the same as the `appendBytes()` method that could be used for streaming content, but it would at least allow some dynamically loaded mp4 video files to be passed in to the `NetStream.play()` method.

To accept MP4 as chunks into the `appendBytes()` method itself would take a bit more effort .. let me check with one of my colleagues on that as I do recollect some stuff going in there to parse MP4 files..

thanks

**Comment by xpojosh** [comment]
@ajwfrost  Thanks so much for diving into this so quickly! 

This is exciting. I'll keep an eye out for the new release.

josh

**Reply by Jason0247** [reply]
Is there a complete code for this part of the video loop that you can refer to? I'm not very familiar with version 51 yet, thank you!

**Reply by xpojosh** [reply]
@Jason0247 -- There are a few ways to skin this cat. The code snippet above in this thread was based on the work found here: [https://blog.redannick.com/seamlessly-looping-video-files-in-as3-with-netstream-appendbytes/](url) -- the big difference is that that example appends the bytes on the onMetadata call, whereas my example here meters things out at a target rate on the onEnterFrame event. 

The approach was available in version prior to 51, but this latest release fixes a bug that was preventing all frames from being rendered.

**Reply by Jason0247** [reply]
Does this method only support FLV?

**Reply by ajwfrost** [reply]
Sending data in bit by bit via `appendBytes` - yes it currently only supports FLV, but there is a feature request for us to add support for the MP4 container too. See #3281
